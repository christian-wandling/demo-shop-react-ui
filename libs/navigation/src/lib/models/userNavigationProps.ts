import { UserResponse } from '@demo-shop-react-ui/api';

export interface UserNavigationProps {
  user: UserResponse | null;
  className: string;
  'data-testid'?: string;
  onLogin: () => void;
  onLogout: () => void;
  onRegister: () => void;
}
