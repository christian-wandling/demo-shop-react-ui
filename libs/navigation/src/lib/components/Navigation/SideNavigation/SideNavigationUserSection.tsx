import { UserNavigationProps } from '../../../models/userNavigationProps';

export const SideNavigationUserSection = ({ user, onLogin, onLogout, onRegister }: UserNavigationProps) => {
  if (user) {
    return (
      <div className="lg:hidden space-y-6 border-t border-gray-200 px-4 py-6">
        <div className="flow-root">
          <a className="-m-2 block p-2 font-medium text-xl text-gray-400">
            {user.firstname}&nbsp;{user.lastname}
          </a>
        </div>
        <div className="flow-root">
          <a
            onClick={onLogout}
            className="-m-2 block p-2 font-medium text-xl text-gray-700 hover:text-indigo-500 cursor-pointer">
            Sign out
          </a>
        </div>
      </div>
    );
  }

  return (
    <div className="lg:hidden space-y-6 border-t border-gray-200 px-4 py-6">
      <div className="flow-root">
        <a
          onClick={onLogin}
          className="-m-2 block p-2 font-medium text-xl text-gray-700 hover:text-indigo-500 cursor-pointer">
          Sign in
        </a>
      </div>
      <div className="flow-root">
        <a
          onClick={onRegister}
          className="-m-2 block p-2 font-medium text-xl text-gray-700 hover:text-indigo-500 cursor-pointer">
          Register
        </a>
      </div>
    </div>
  );
};
