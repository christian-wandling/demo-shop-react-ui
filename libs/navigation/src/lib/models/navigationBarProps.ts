import { NavigationItem } from './navigationItem';
import { UserResponse } from '@demo-shop-react-ui/api';

export interface NavigationBarProps {
  items: NavigationItem[];
  selectedItem: string;
  user?: UserResponse;
  onOpenSideNavigation: () => void;
  onSelectedItemChange: (item: string) => void;
  onLogin: () => void;
  onRegister: () => void;
  onLogout: () => void;
}
