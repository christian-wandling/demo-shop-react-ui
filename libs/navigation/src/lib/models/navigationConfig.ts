import { NavigationItem } from './navigationItem';

export interface NavigationConfig {
  items: NavigationItem[];
}
