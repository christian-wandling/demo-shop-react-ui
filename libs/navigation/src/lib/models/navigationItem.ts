import { PermissionStrategy } from '@demo-shop/auth';
import { NavigationItemType } from '../enums/navigationItemType';
import { RouteItem } from './routeItem';

/**
 * Base abstract class representing a navigation item in the application
 */
export abstract class NavigationItem {
  /**
   * The type of navigation item
   */
  readonly type!: NavigationItemType;

  /**
   * Creates a new navigation item
   * @param {string} label - Display text for the navigation item
   * @param {number} order - Numeric value determining the display order of the item
   * @param {object} [options] - Optional configuration options
   * @param {PermissionStrategy} [options.permissionStrategy] - Strategy for determining user permissions for this item
   * @param {RouteItem[]} [subItems] - Child navigation items nested under this item
   */
  protected constructor(
    public label: string,
    public order: number,
    public options?: {
      permissionStrategy?: PermissionStrategy;
    },
    public subItems?: RouteItem[]
  ) {}
}
