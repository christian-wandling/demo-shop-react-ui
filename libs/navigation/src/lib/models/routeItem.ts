import { NavigationItemType } from '../enums/navigationItemType';
import { NavigationItem } from './navigationItem';
import { PermissionStrategy } from '@demo-shop-react-ui/auth';

/**
 * Represents a navigation item that links to an Angular route
 */
export class RouteItem extends NavigationItem {
  /**
   * The type of navigation item, always set to ROUTE for RouteItem
   */
  override readonly type: NavigationItemType = NavigationItemType.ROUTE;

  constructor(
    label: string,
    order: number,
    public override options: {
      route: string;
      permissionStrategy?: PermissionStrategy;
      query?: string;
    },
    subItems?: RouteItem[]
  ) {
    super(label, order, options, subItems);
  }
}
