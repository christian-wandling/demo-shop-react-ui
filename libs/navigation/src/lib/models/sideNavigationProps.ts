import { NavigationItem } from './navigationItem';
import { UserResponse } from '@demo-shop-react-ui/api';

export interface SideNavigationProps {
  items: NavigationItem[];
  selectedItem: string;
  user: UserResponse | null;
  isOpen: boolean;
  onSetOpen: (isOpen: boolean) => void;
  onSelectedItemChange: (item: string) => void;
  onLogin: () => void;
  onRegister: () => void;
  onLogout: () => void;
}
