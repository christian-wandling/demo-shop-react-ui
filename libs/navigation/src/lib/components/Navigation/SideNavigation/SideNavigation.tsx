import { Link } from 'react-router';
import { AnimatePresence, motion } from 'framer-motion';
import { SideNavigationProps } from '../../../models/sideNavigationProps';
import { RouteItem } from '../../../models/routeItem';
import { SideNavigationUserSection } from './SideNavigationUserSection';

/**
 * NavigationMobile renders the mobile version of the navigation.
 * It includes a slide-over menu with links and user navigation.
 */
export const SideNavigation = ({
  user,
  items,
  selectedItem,
  isOpen,
  onSetOpen,
  onSelectedItemChange,
  onLogin,
  onRegister,
  onLogout,
}: SideNavigationProps) => {
  return (
    <AnimatePresence>
      {isOpen && (
        <>
          <motion.dialog
            initial={{ opacity: 0 }}
            animate={{ opacity: 1 }}
            exit={{ opacity: 0 }}
            onClick={() => onSetOpen(false)}
            onKeyDown={e => e.key === 'Escape' && onSetOpen(false)}
            className="relative flex z-40 lg:hidden"
            aria-modal="true">
            <div className="fixed inset-0 bg-black bg-opacity-25" aria-hidden="true" data-testid="dialog"></div>
          </motion.dialog>

          <motion.div
            initial={{ x: '-100%' }}
            animate={{ x: 0 }}
            exit={{ x: '-100%' }}
            transition={{ type: 'tween' }}
            className="fixed h-screen top-0 z-40 flex w-full max-w-xs flex-col overflow-y-auto bg-white px-6 p-2 shadow-xl">
            <div className="sticky z-20 top-0 flex px-4 pb-2 pt-5">
              <button
                onClick={() => onSetOpen(false)}
                type="button"
                className="relative opacity-90 -m-2 inline-flex items-center justify-center rounded-md p-2 text-gray-400 bg-white">
                <span className="absolute -inset-0.5"></span>
                <span className="sr-only">Close menu</span>
                <svg
                  className="h-6 w-6"
                  fill="none"
                  viewBox="0 0 24 24"
                  strokeWidth="1.5"
                  stroke="currentColor"
                  aria-hidden="true">
                  <path strokeLinecap="round" strokeLinejoin="round" d="M6 18L18 6M6 6l12 12" />
                </svg>
              </button>
            </div>

            <div className="flex-1 space-y-6 border-t border-gray-200 px-4 py-6">
              {items.map((item, index) => (
                <div key={index} className="flow-root">
                  <Link
                    to={(item as RouteItem).options?.route}
                    onClick={() => {
                      onSetOpen(false);
                      onSelectedItemChange(item.label);
                    }}
                    className={
                      (selectedItem === item.label ? 'text-indigo-600' : 'text-gray-700') +
                      '-m-2 block p-2 font-medium text-gray-700 hover:text-indigo-500 text-xl'
                    }>
                    {item.label.charAt(0).toUpperCase() + item.label.slice(1)}
                  </Link>
                </div>
              ))}
            </div>

            <SideNavigationUserSection user={user} onLogin={onLogin} onRegister={onRegister} onLogout={onLogout} />
          </motion.div>
        </>
      )}
    </AnimatePresence>
  );
};
