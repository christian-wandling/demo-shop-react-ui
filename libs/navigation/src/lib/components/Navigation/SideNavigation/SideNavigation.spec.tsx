import { beforeEach, describe, expect, it, vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import { SideNavigation } from './SideNavigation';
import { MemoryRouter } from 'react-router';
import { SideNavigationProps } from '../../../models/sideNavigationProps';
import { RouteItem } from '../../../models/routeItem';

vi.mock('framer-motion', () => ({
  motion: {
    dialog: ({ children, onClick, onKeyDown, className }: any) => (
      <dialog onClick={onClick} onKeyDown={onKeyDown} className={className}>
        {children}
      </dialog>
    ),
    div: ({ children, className }: any) => <div className={className}>{children}</div>,
  },
  AnimatePresence: ({ children }: any) => <>{children}</>,
}));

vi.mock('../UserNavigation/UserNavigation', () => ({
  UserNavigation: ({ user }: any) => (
    <div data-testid="navigation-slide-over-user-section">{user ? `Logged in as ${user.name}` : 'Not logged in'}</div>
  ),
}));

describe('SideNavigation', () => {
  const mockProps: SideNavigationProps = {
    user: {
      email: 'test@example.com',
      id: 0,
      firstname: 'firstname',
      lastname: 'lastname',
      phone: null,
    },
    items: [new RouteItem('products', 101, { route: 'products' }), new RouteItem('orders', 102, { route: 'orders' })],
    selectedItem: 'products',
    isOpen: true,
    onSetOpen: vi.fn(),
    onSelectedItemChange: vi.fn(),
    onLogin: vi.fn(),
    onRegister: vi.fn(),
    onLogout: vi.fn(),
  };

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders when isOpen is true', () => {
    render(
      <MemoryRouter>
        <SideNavigation {...mockProps} />
      </MemoryRouter>
    );

    expect(screen.getByText('Products')).toBeTruthy();
    expect(screen.getByText('Orders')).toBeTruthy();
    expect(screen.getByTestId('navigation-slide-over-user-section')).toBeTruthy();
  });

  it('does not render when isOpen is false', () => {
    render(
      <MemoryRouter>
        <SideNavigation {...mockProps} isOpen={false} />
      </MemoryRouter>
    );

    expect(screen.queryByText('Products')).not.toBeTruthy();
    expect(screen.queryByText('Orders')).not.toBeTruthy();
  });

  it('calls onSetOpen with false when close button is clicked', () => {
    render(
      <MemoryRouter>
        <SideNavigation {...mockProps} />
      </MemoryRouter>
    );

    fireEvent.click(screen.getByRole('button'));
    expect(mockProps.onSetOpen).toHaveBeenCalledWith(false);
  });

  it('calls onSetOpen with false when backdrop is clicked', () => {
    render(
      <MemoryRouter>
        <SideNavigation {...mockProps} />
      </MemoryRouter>
    );

    const backdrop = screen.getByTestId('dialog');
    fireEvent.click(backdrop);

    expect(mockProps.onSetOpen).toHaveBeenCalledWith(false);
  });

  it('applies active styling to selected item', () => {
    render(
      <MemoryRouter>
        <SideNavigation {...mockProps} />
      </MemoryRouter>
    );

    const homeLink = screen.getByText('Products');
    const aboutLink = screen.getByText('Orders');

    expect(homeLink.className).toContain('text-indigo-600');
    expect(aboutLink.className).not.toContain('text-indigo-600');
  });

  it('calls onSelectedItemChange and onSetOpen when a navigation item is clicked', () => {
    render(
      <MemoryRouter>
        <SideNavigation {...mockProps} />
      </MemoryRouter>
    );

    fireEvent.click(screen.getByText('Orders'));

    expect(mockProps.onSelectedItemChange).toHaveBeenCalledWith('orders');
    expect(mockProps.onSetOpen).toHaveBeenCalledWith(false);
  });

  describe('Snapshots', () => {
    it('matches snapshot when open', () => {
      const { container } = render(
        <MemoryRouter>
          <SideNavigation {...mockProps} />
        </MemoryRouter>
      );
      expect(container).toMatchSnapshot();
    });

    it('matches snapshot when closed', () => {
      const { container } = render(
        <MemoryRouter>
          <SideNavigation {...mockProps} isOpen={false} />
        </MemoryRouter>
      );
      expect(container).toMatchSnapshot();
    });

    it('matches snapshot without a user (logged out state)', () => {
      const propsWithoutUser = { ...mockProps, user: null };

      const { container } = render(
        <MemoryRouter>
          <SideNavigation {...propsWithoutUser} />
        </MemoryRouter>
      );
      expect(container).toMatchSnapshot();
    });
  });
});
