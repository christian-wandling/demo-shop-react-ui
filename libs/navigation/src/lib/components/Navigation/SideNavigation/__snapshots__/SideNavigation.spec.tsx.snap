// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SideNavigation > Snapshots > matches snapshot when closed 1`] = `<div />`;

exports[`SideNavigation > Snapshots > matches snapshot when open 1`] = `
<div>
  <dialog
    class="relative flex z-40 lg:hidden"
  >
    <div
      aria-hidden="true"
      class="fixed inset-0 bg-black bg-opacity-25"
      data-testid="dialog"
    />
  </dialog>
  <div
    class="fixed h-screen top-0 z-40 flex w-full max-w-xs flex-col overflow-y-auto bg-white px-6 p-2 shadow-xl"
  >
    <div
      class="sticky z-20 top-0 flex px-4 pb-2 pt-5"
    >
      <button
        class="relative opacity-90 -m-2 inline-flex items-center justify-center rounded-md p-2 text-gray-400 bg-white"
        type="button"
      >
        <span
          class="absolute -inset-0.5"
        />
        <span
          class="sr-only"
        >
          Close menu
        </span>
        <svg
          aria-hidden="true"
          class="h-6 w-6"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 18L18 6M6 6l12 12"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
    </div>
    <div
      class="flex-1 space-y-6 border-t border-gray-200 px-4 py-6"
    >
      <div
        class="flow-root"
      >
        <a
          class="text-indigo-600-m-2 block p-2 font-medium text-gray-700 hover:text-indigo-500 text-xl"
          data-discover="true"
          href="/products"
        >
          Products
        </a>
      </div>
      <div
        class="flow-root"
      >
        <a
          class="text-gray-700-m-2 block p-2 font-medium text-gray-700 hover:text-indigo-500 text-xl"
          data-discover="true"
          href="/orders"
        >
          Orders
        </a>
      </div>
    </div>
    <div
      data-testid="user-section"
    >
      Logged in as undefined
    </div>
  </div>
</div>
`;

exports[`SideNavigation > Snapshots > matches snapshot without a user (logged out state) 1`] = `
<div>
  <dialog
    class="relative flex z-40 lg:hidden"
  >
    <div
      aria-hidden="true"
      class="fixed inset-0 bg-black bg-opacity-25"
      data-testid="dialog"
    />
  </dialog>
  <div
    class="fixed h-screen top-0 z-40 flex w-full max-w-xs flex-col overflow-y-auto bg-white px-6 p-2 shadow-xl"
  >
    <div
      class="sticky z-20 top-0 flex px-4 pb-2 pt-5"
    >
      <button
        class="relative opacity-90 -m-2 inline-flex items-center justify-center rounded-md p-2 text-gray-400 bg-white"
        type="button"
      >
        <span
          class="absolute -inset-0.5"
        />
        <span
          class="sr-only"
        >
          Close menu
        </span>
        <svg
          aria-hidden="true"
          class="h-6 w-6"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 18L18 6M6 6l12 12"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
    </div>
    <div
      class="flex-1 space-y-6 border-t border-gray-200 px-4 py-6"
    >
      <div
        class="flow-root"
      >
        <a
          class="text-indigo-600-m-2 block p-2 font-medium text-gray-700 hover:text-indigo-500 text-xl"
          data-discover="true"
          href="/products"
        >
          Products
        </a>
      </div>
      <div
        class="flow-root"
      >
        <a
          class="text-gray-700-m-2 block p-2 font-medium text-gray-700 hover:text-indigo-500 text-xl"
          data-discover="true"
          href="/orders"
        >
          Orders
        </a>
      </div>
    </div>
    <div
      data-testid="user-section"
    >
      Not logged in
    </div>
  </div>
</div>
`;
